src/System.Private.CoreLib/shared/System/IO/Path.Windows.cs

```diff
@@ -252,24 +252,21 @@ internal static ReadOnlySpan<char> GetVolumeName(ReadOnlySpan<char> path)
             {
                 // Cut from "\\?\UNC\Server\Share" to "Server\Share"
                 // Cut from  "\\Server\Share" to "Server\Share"
-                return TrimEndingDirectorySeparator(root.Slice(offset));
+                return TrimEndingDirectorySeparatorHelper(root.Slice(offset));
             }
             else if (PathInternal.IsDevice(path))
             {
-                return TrimEndingDirectorySeparator(root.Slice(4)); // Cut from "\\?\C:\" to "C:"
+                return TrimEndingDirectorySeparatorHelper(root.Slice(4)); // Cut from "\\?\C:\" to "C:"
             }
 
-            return TrimEndingDirectorySeparator(root); // e.g. "C:"
-        }
+            return TrimEndingDirectorySeparatorHelper(root); // e.g. "C:"
 
-        /// <summary>
-        /// Trims the ending directory separator if present.
-        /// </summary>
-        /// <param name="path"></param>
-        internal static ReadOnlySpan<char> TrimEndingDirectorySeparator(ReadOnlySpan<char> path) =>
-            PathInternal.EndsInDirectorySeparator(path) ?
-                path.Slice(0, path.Length - 1) :
-                path;
+            // Trim ending directory separator
+            ReadOnlySpan<char> TrimEndingDirectorySeparatorHelper(ReadOnlySpan<char> pathToTrim) =>
```

Comment:
Why can't we use the "real" one?  Does the extra !IsRoot check mess something up?

Reply:
> Does the extra !IsRoot check mess something up?

Yes some of this tests fail https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Extensions/tests/System/IO/PathTests_Windows.netcoreapp.cs#L255

Reply:
Also on comment `// e.g. "C:"`

Reply:
If we are not able to use the API that we are exposing ourselves, it is usually a red flag that the API is broken.

@JeremyKuhne @cdmihai What do we expect this API to be used for? Are there examples in msbuild where this API will be used?

Reply:
I don't know if my fault...I worked to expose APIs reported on issue https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/IO/PathInternal.cs#L27-L35 and after that I found another API with same name as helper for `GetVolumeName` API.

Reply:
Here are the msbuild path helpers that motivated adding this API: https://github.com/Microsoft/msbuild/blob/master/src/Shared/FileUtilities.cs#L192

The current msbuild logic also using the "unique" logic that treats `c:\` as ending with slash and stripping it as necessary. 

There is whole family of these helpers. The most popular one is `EnsureTrailingSlash` (30+ hits including `WithTrailingSlash` convenience helper). `EnsureNoTrailingSlash` that we are adding here has 12 hits.

If we were to pick one API from this family, should it be the most popular one (ie `EnsureTrailingSlash`)?

It would match the msbuild public property functions: There is `EnsureTrailingSlash` [property function](https://docs.microsoft.com/en-us/visualstudio/msbuild/property-functions), but none of the other ones.

Reply:
MSBuild does a lot of path comparisons, and so it needs to normalize the paths first (normalize slashes, eliminate relative path fragments, etc). Eliminating the trailing slash via a span would eliminate some allocations. Maybe a better approach would be in memory complicated comparisons as opposed to "dumb-yet-allocating" normalizations, or better yet, implementing abstract path types instead of using strings, but that's a different conversation.

Reply:
> MSBuild does a lot of path comparison. 
> Maybe a better approach would be in memory complicated comparison

Agree. If the actual problem that this is trying to solve is efficient non-allocating comparison of two paths with normalization, it would be better to have an API for that, not an API that solves 2% of the problem.

Reply:
> If we were to pick one API from this family, should it be the most popular one (ie EnsureTrailingSlash)?

API review shot it down, but I don't agree with that decision- particularly with the extra evidence you have.

> it would be better to have an API for that, not an API that solves 2% of the problem.

The fundamental trick here is that it is extremely difficult to write such an API that will be understandable and/or have people use it correctly. Do you consider actual equivalence? Case sensitivity alone is a hard nut to crack as it can vary from segment to segment along the path. Alternative volume syntaxes on Windows makes this really hard to do as well. What about paths that actually point to the same file (sym/hard links)? It is a pretty common thing to assume that if paths aren't the same then they must point to different files. While that usually doesn't cause problems, it can. I look at `Path.IsPathRooted` as an example for how bad assumptions go awry.

I think the least we can do in the short term is ease the composing of path logic in the short term by eliminating the need to deal with multiple separator chars and avoid the common mistake of over trimming. We use these methods all over internally and anyone who builds additional logic ends up doing the same.

Reply:
True, path comparison kind of depends on the application. MSBuild for example does not care about links when comparing paths. Maybe a simple .net path comparison API could just call GetFullPath (don't remember from the top of my head whether this one normalizes the path) and perform the comparison on the native buffers without converting them to .net objects.

Reply:
@JeremyKuhne I'm confused as to the result of this discussion, is it to continue, or to return to API review?

Reply:
I'll make the request to add back the other API and bring it back to review.

Reply:
I think this one should go in as I don't think we'll remove this API.